bullet-featherstone/src/KinematicsFeatures.cc

```diff
@@ -59,60 +69,55 @@ FrameData3d KinematicsFeatures::FrameDataRelativeToWorld(
         model = this->ReferenceInterface<ModelInfo>(linkInfo2->model);
-
-        if (indexOpt2.has_value())
+        if (linkInfo2->indexInModel.has_value())
         {
-          const auto index2 = *indexOpt2;
-          FrameData data;
-          data.pose = GetWorldTransformOfLink(*model, *linkInfo2);
-          const auto &link = model->body->getLink(index2);
-          data.linearVelocity = convert(
-            link.m_absFrameTotVelocity.getLinear());
-          data.angularVelocity = convert(
-            link.m_absFrameTotVelocity.getAngular());
-          return data;
+          return getNonBaseLinkFrameData(model, linkInfo2.get());
         }
       }
     }
-
-    auto collisionIt = this->collisions.find(_id.ID());
-    if (collisionIt != this->collisions.end())
+    else
     {
-      const auto &collisionInfo = collisionIt->second;
-
-      const auto linkIt2 = this->links.find(collisionInfo->link);
-      if (linkIt2 != this->links.end())
+      auto collisionIt = this->collisions.find(_id.ID());
+      if (collisionIt != this->collisions.end())
       {
-        const auto &linkInfo2 = linkIt2->second;
-        const auto indexOpt2 = linkInfo2->indexInModel;
-        model = this->ReferenceInterface<ModelInfo>(linkInfo2->model);
+        isCollision = true;
+        const auto &collisionInfo = collisionIt->second;
 
-        if (indexOpt2.has_value())
+        const auto linkIt2 = this->links.find(collisionInfo->link);
+        if (linkIt2 != this->links.end())
         {
-          const auto index2 = *indexOpt2;
-          FrameData data;
-          data.pose = GetWorldTransformOfLink(*model, *linkInfo2);
-          const auto &link = model->body->getLink(index2);
-          data.linearVelocity = convert(
-            link.m_absFrameTotVelocity.getLinear());
-          data.angularVelocity = convert(
-            link.m_absFrameTotVelocity.getAngular());
-          return data;
+          const auto &linkInfo2 = linkIt2->second;
+          model = this->ReferenceInterface<ModelInfo>(linkInfo2->model);
+          collisionPoseOffset = collisionInfo->linkToCollision;
+          if (linkInfo2->indexInModel.has_value())
+          {
+            auto data = getNonBaseLinkFrameData(model, linkInfo2.get());
+            data.pose = data.pose * collisionPoseOffset;
+            return data;
+          }
+        }
+      }
+      else
+      {
+        auto modelIt = this->models.find(_id.ID());
+        if (modelIt != this->models.end())
+        {
+          model = modelIt->second.get();
+          isModel = true;
         }
       }
     }
-
-    if (!model || model->body == nullptr)
-      model = this->FrameInterface<ModelInfo>(_id);
   }
 
   FrameData data;
-  if(model && model->body)
+  if (model && model->body)
   {
-    data.pose = convert(model->body->getBaseWorldTransform())
-      * model->baseInertiaToLinkFrame;
+    data.pose = convert(model->body->getBaseWorldTransform());
```

Comment:
Don't we have to account for 
- The offset from the model to the base link, i.e, the links pose in SDF? Do we have a test that covers this?
- The offset of the link frame to the center of mass? I thought that's what `model->baseInertiaToLinkFrame` was there for .

Reply:
This line computes the pose for the model. The next 2 lines takes into account the pose of the link (and inertia offset if any) which uses `model->baseInertiaToLinkFrame`. I think that's correct? I added a test in [5c6ec0f](https://github.com/gazebosim/gz-physics/pull/647/commits/5c6ec0ff7adc80245b0653580c96cf306a48a7bc) that checks the frame data for a link and a collision that has a pose offset.

Reply:
expanded test to cover both base and non-base links. [84054fa](https://github.com/gazebosim/gz-physics/pull/647/commits/84054faf625898319f469f3f6eff0a11d18f238d)

Reply:
It looks like this test is not testing model frame semantics at all. Fortunately, we currently don't use it in gz-sim, but it would be good to fix (not in this PR). I've added a failing test in the `azeey/kinematic_feature_failing_test`  branch. Evidently, dartsim fails the test as well.